### PR TITLE
[3.0] Clean Form, Validator, DowCrawler and some more

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Form/FormFactory.php [hunk 1]
```diff
@@ -61,36 +61,12 @@ public function createForProperty($class, $property, $data = null, array $option
      */
     public function createBuilder($type = 'Symfony\Component\Form\Extension\Core\Type\FormType', $data = null, array $options = array())
     {
-        $name = null;
-        $typeName = null;
-
         if ($type instanceof ResolvedFormTypeInterface) {
```
Comment (Tobion) [inline src/Symfony/Component/Form/FormFactory.php]: getBlockPrefix must be added to ResolvedFormTypeInterface

Comment (Tobion) [inline src/Symfony/Component/Form/FormFactory.php]: I'm not sure we actually need to keep support for `ResolvedFormTypeInterface` in the factory. it is not supported according to FormFactoryInterface

Comment (Tobion) [inline src/Symfony/Component/Form/FormFactory.php]: I think there is a bug in the current implementation. createBuilder has logic to handle ResolvedFormTypeInterface, but it then calls createNamedBuilder which will throw an exception as it requires a string as the type.

Comment (Tobion) [inline src/Symfony/Component/Form/FormFactory.php]: Yeah, it must be removed. It was deprecated in 2.8: https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Form/FormFactory.php#L114
So it must be removed from this method as well.

Comment (Tobion) [inline src/Symfony/Component/Form/FormFactory.php]: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/FormFactoryInterface.php must be updated as well as it still says `string|FormTypeInterface $type`

Comment (nicolas-grekas) [inline src/Symfony/Component/Form/FormFactory.php]: Updated, can you please review again?

File: src/Symfony/Component/Form/FormFactoryInterface.php [hunk 2]
```diff
@@ -21,9 +21,9 @@
      *
      * @see createBuilder()
      *
-     * @param string|FormTypeInterface $type    The type of the form
-     * @param mixed                    $data    The initial data
-     * @param array                    $options The options
+     * @param string $type    The type of the form
```
Comment (Tobion) [inline src/Symfony/Component/Form/FormFactoryInterface.php]: there are more of it in the interface

Comment (nicolas-grekas) [inline src/Symfony/Component/Form/FormFactoryInterface.php]: fixed

File: src/Symfony/Component/Form/Extension/Validator/EventListener/ValidationListener.php [hunk 3]
```diff
@@ -35,10 +35,6 @@ public static function getSubscribedEvents()
         return array(FormEvents::POST_SUBMIT => 'validateForm');
     }
 
-    /**
-     * @param ValidatorInterface       $validator
-     * @param ViolationMapperInterface $violationMapper
-     */
```
Comment (harikt) [inline src/Symfony/Component/Form/Extension/Validator/EventListener/ValidationListener.php]: May I know what is the reason of removing the docblock comments ?

Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Validator/EventListener/ValidationListener.php]: they provide no value at all, this is just an exact repetition of the method signature

Comment (harikt) [inline src/Symfony/Component/Form/Extension/Validator/EventListener/ValidationListener.php]: ok, thanks. I wonder whether this have any impact on the sami / phpdoc .

**Comment by Tobion** [comment]
Please also remove max_length from https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/FormFactoryInterface.php#L96

**Comment by nicolas-grekas** [comment]
Removed

**Comment by Tobion** [comment]
The `choice_list` option must be removed still I think.

**Comment by nicolas-grekas** [comment]
@Tobion I agree with you, but this is more tricky and could be left for an other PR.
I added a few more removals btw.

**Comment by fabpot** [comment]
The `choice_list` removal should be done as its own PR.

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 4]
```diff
@@ -846,136 +846,6 @@ public static function xpathLiteral($s)
     }
 
     /**
-     * @deprecated Using the SplObjectStorage API on the Crawler is deprecated as of 2.8 and will be removed in 3.0.
-     */
-    public function attach($object, $data = null)
```
Comment (Tobion) [inline src/Symfony/Component/DomCrawler/Crawler.php]: removing this without actually removing `extends \SplObjectStorage` makes no sense

**Comment by Tobion** [comment]
Apart from the changes in `Crawler` :+1:

**Comment by nicolas-grekas** [comment]
Crawler updated

File: src/Symfony/Component/Debug/ExceptionHandler.php [hunk 5]
```diff
@@ -38,12 +38,6 @@ class ExceptionHandler
 
     public function __construct($debug = true, $charset = null, $fileLinkFormat = null)
     {
-        if (false !== strpos($charset, '%')) {
-            // Swap $charset and $fileLinkFormat for BC reasons
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: See #16085

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by renatomefidf** [comment]
Hello @nicolas-grekas 
I see you removed the support for createForm using objects:
https://github.com/symfony/symfony/commit/abca2d6fdc89479586e052cd11241d5a9bc885e5#diff-5c1348d69be32426ff20446c9e812365R64

How can I create forms which comes from a service for instance?

**Comment by stof** [comment]
@renatomefidf defining form types as service **never** required passing the instance to createForm. If you are passing an instance here, it means you are **not** using your form type as a service as far as the Form component is concerned.
When the form type is registered as a service, using it only involves using its name (see how you use the Symfony core types: they are defined as services)

**Comment by renatomefidf** [comment]
Thank you @stof 
Its actually not a problem I'm facing, but a question in here:
http://stackoverflow.com/questions/33451498/symfony-3-define-form-as-a-service

Thanks again for helping and making it clearer!

**Comment by stof** [comment]
OK, I answered on SO itself

**Comment by PaddyLock** [comment]
@stof I am trying to do the same as @renatomefidf But I can't get it to work. Should I be using the form name defined in the form class within getBlockPrefix(), or the forms service name or the name defined in the service definition within 
`tags:
    - {name: 'form.my_name'}`

I have tried with

`$this->createForm('form.my_name', $entity);`
`$this->createForm('my_form_service_name', $entity);`
`$this->createForm(MyFormType::class, $entity);`

The last one uses the form class but not the form as a service, which is what I need.

Thanks

**Comment by xabbuh** [comment]
@PaddyLock The tag name to add a form type service is `form.type`. But please use one of [the support channels](http://symfony.com/support) if you need further helper. Thank you!

**Pull request merged**